File: servers/server_clear_thought/tools/safe_struggle_designer.py
from __future__ import annotations

from typing import Any, Dict, List

from pydantic import BaseModel

from ..core.base_tool import BaseTool


class SafeStruggleDesigner(BaseTool):
    slug = "safe_struggle_designer"
    endpoint_path = "/safe-struggle-designer"

    class InputSchema(BaseModel):
        skill: str
        current_level: int
        target_level: int
        constraints: Dict[str, Any] | None = None

    class OutputSchema(BaseModel):
        scaffold_steps: List[str]
        safety_measures: List[str]
        review_intervals: str

    def execute(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        steps = [
            f"Practice {payload['skill']} at level {lvl}"
            for lvl in range(payload["current_level"], payload["target_level"] + 1)
        ]
        measures = ["Take breaks", "Monitor progress"]
        return {
            "scaffold_steps": steps,
            "safety_measures": measures,
            "review_intervals": "weekly",
        }

File: servers/server-clear-thought/src/tools/safe-struggle-designer.ts
import { z } from 'zod';
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import type { SessionState } from '../state/SessionState.js';

export function registerSafeStruggleDesigner(server: McpServer, _sessionState: SessionState) {
  server.tool(
    'safe_struggle_designer',
    'Design a safe skill improvement plan',
    {
      skill: z.string(),
      current_level: z.number(),
      target_level: z.number(),
      constraints: z.record(z.any()).optional()
    },
    async ({ skill, current_level, target_level }) => {
      const scaffold_steps = [] as string[];
      for (let lvl = current_level; lvl <= target_level; lvl++) {
        scaffold_steps.push(`Practice ${skill} at level ${lvl}`);
      }
      const safety_measures = ['Take breaks', 'Monitor progress'];
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify({ scaffold_steps, safety_measures, review_intervals: 'weekly' }, null, 2)
          }
        ]
      };
    }
  );
}
